docs/source/common_issues.rst

```diff
@@ -339,6 +339,24 @@ style anyway).  We can write the above code without a cast by using
            g(o + 1)        # Okay; type of o is inferred as int here
            ...
 
+Note that MyPy cannot always narrow a type based on ``isinstance()``
+used within a ternary/conditional expression, e.g.:
+
+.. code-block:: python
+
+	def f(x: AnyStr, y: AnyStr) -> bytes:
```

Comment:
The example with `AnyStr` is not the best one, since it combines two tricky points:
* the one you are explaining
* the fact that functions generic in variables with restrictions are type checked multiple times.

I would rather use a single argument function with a `Union[str, bytes]` argument.

Reply:
Actually, if I try to simplify this down to ```def f(x: Union[str, bytes]) -> bytes:
    r = x.encode('ascii') if isinstance(x, str) else x
    # results in 'error: "bytes" has no attribute "encode"; maybe "decode"?'

    if isinstance(x, str):
        r = x.encode('ascii')  # whereas Mypy can narrow within an if statement
    else:
        r = x
    return r```

Mypy doesn't error at all, which is quite interesting. Whereas if I use AnyStr on that same argument, there is an error.

It may be that I have oversimplified the explanation for what is going on?

Reply:
To clarify, mypy _can_ restrict types using `isinstance()` in ternary expressions, but it cannot consider sub-expressions of ternary expression _unreachable_. Here is even simpler example:
```python
x: bytes

if isinstance(x, str):
    r = x.encode()  # OK, this is unreachable
else:
    r = x

r = x.encode() if isinstance(x, str) else x  # Error here
```